GitHubReferences.ps1

```diff
@@ -0,0 +1,227 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+function Get-GitHubReference
+{
+<#
+    .SYNOPSIS
+        Retrieve a reference of a given GitHub repository.
+
+    .DESCRIPTION
+        Retrieve a reference of a given GitHub repository.
+        The Git repo for this module can be found here: http://aka.ms/PowerShellForGitHub
+
+    .PARAMETER OwnerName
+        Owner of the repository.
+        If not supplied here, the DefaultOwnerName configuration property value will be used.
+
+    .PARAMETER RepositoryName
+        Name of the repository.
+        If not supplied here, the DefaultRepositoryName configuration property value will be used.
+
+    .PARAMETER Uri
+        Uri for the repository.
+        The OwnerName and RepositoryName will be extracted from here instead of needing to provide
+        them individually.
+
+    .PARAMETER Reference
+        Name of the reference, for example: "heads/<branch name>" for branches and "tags/<tag name>" for tags
+
+    .PARAMETER AccessToken
+        If provided, this will be used as the AccessToken for authentication with the
+        REST Api.  Otherwise, will attempt to use the configured value or will run unauthenticated.
+
+    .PARAMETER NoStatus
+        If this switch is specified, long-running commands will run on the main thread
+        with no commandline status update.  When not specified, those commands run in
+        the background, enabling the command prompt to provide status information.
+        If not supplied here, the DefaultNoStatus configuration property value will be used.
+    
+    .OUTPUTS
+        [PSCustomObject]
+        Details of the git reference in the given repository
+
+    .EXAMPLE
+        Get-GitHubReference -OwnerName Powershell -RepositoryName PowerShellForGitHub -Reference heads/master
+#>
+    [CmdletBinding(
+        SupportsShouldProcess,
+        DefaultParametersetName='Elements')]
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSShouldProcess", "", Justification="Methods called within here make use of PSShouldProcess, and the switch is passed on to them inherently.")]
+    param(
+        [Parameter(ParameterSetName='Elements')]
+        [string] $OwnerName,
+
+        [Parameter(ParameterSetName='Elements')]
+        [string] $RepositoryName,
+
+        [Parameter(
+            Mandatory,
+            ParameterSetName='Uri')]
+        [string] $Uri,
+
+        [Parameter(Mandatory)]
+        [string] $Reference,

```

Comment:
If you make this optional (leaving off `[Parameter(Mandatory)]`, then this will handle _both_ a [single reference](https://developer.github.com/v3/git/refs/#get-a-reference) and [all references](https://developer.github.com/v3/git/refs/#get-all-references), but you'd also have to call `Invoke-GHRestMethodMultipleResult` and branch on what you set `$description` to.  This is a common pattern you can see in other methods (like `Get-GitHubIssue`)

Reply:
Thanks for the note, I'll have a look at the other methods and adapt the same approach here

Reply:
@HowardWolosky looks like the API/docs have been updated and the single/all reference links are no longer available. Can I understand getting a reference to mean _getting a single reference_ and getting all references to mean _listing matching references_? In that case should I create a new method for listing matching references?

Reply:
Looks like it would now be [List Matching References](https://developer.github.com/v3/git/refs/#list-matching-references).

Reply:
Some thought needs to go in to how you would make a combined single/multiple function given the additional suggestion of splitting out `$Reference` as discussed below.